enhancements/cloud-integration/openstack/cloud-provider-config-upgrade.md

```diff
@@ -0,0 +1,327 @@
+will become the default for OpenStack-based clusters in 4.11.
+
+#### Tech Preview -> GA
+
+For GA we need test test coverage of upgrade and fresh install.
+
+- Existing E2E upgrade and fresh install tests passing
+- The feature will land early in the 4.11 development cycle
+- User facing documentation is updated to highlight configuration options which
+  are deprecated between 4.10 and 4.11
+- User facing documentation is updated to show the new configuration
+
+#### Removing a deprecated feature
+
+N/A. This is a new feature.
+
+### Upgrade / Downgrade Strategy
+
+N/A: Downgrade from CCM is not supported
+
+### Version Skew Strategy
+
+Version skew is handled by maintaining the legacy configuration while writing
+the new configuration to a new configmap. This strategy does not require any
+changes to the existing external CCM upgrade workflow: legacy components
+continue to read the legacy configuration until they are restarted, and the new
+component can read the new configuration at any time.
+
+### Operational Aspects of API Extensions
+
+N/A
+
+#### Failure Modes
+
+In general, failure of CCO will not have a critical impact on the cluster unless
+the user has also manually deleted the ConfigMaps it creates in
+`openshift-config-managed`.
+
+A failure of CCO will cause updates to user-managed config to not be reflected
+in the cluster's deployed configuration. However, the cluster will continue to
+use the previous configuration until the failure of CCO is resolved.
+
+#### Support Procedures
+
+If CCO has failed it will be evident because it will be marked as degraded. If
+running 4.10 this will prevent an upgrade to 4.11. If running 4.11, a valid
+`kube-cloud-config-external` will already exist, but will not be updated to
+reflect new changes to `openshift-config/cloud-provider-config`.
+
+It should not be possible to upgrade a cluster to use external CCM without CCO
+having previously written `kube-cloud-config-external`. However, if it did
+happen, the root cause should be evident because:
+
+- CCO will be marked as degraded
+- CCCMO will be marked as degraded because CCM cannot start (XXX: verify this)
+- Examining the CCM pod will reveal the missing configmap
+
+If the user updates `openshift-config/cloud-provider-config` with incompatible
+configuration, CCO will be marked as degraded and a user-targetted error will
+detail the cause.
+
+## Implementation History
+
+- [2022-01-25] Initial enhancement proposal
+
+## Drawbacks
+
+- The second configmap is ugly and will have to be deprecated at some point in
+  the future.
+
+## Alternatives
+
+- We could reuse `openshift-config-managed/kube-cloud-config` for CCM. This
+  would result in less noise but it would make the upgrade less robust.
+
+- Instead of using CCO, we could write the configmap in a sidecar container to
+  CCM. This is what Azure is doing.
+
+- We could create an OpenStack-specific CRD describing all aspects of cloud
```

Comment:
This would be cool, but the upgrade path would be tricky to navigate I think, worth mentioning the drawbacks of each of the alternatives with them btw, right now I'm thinking some of these sound appealing, tell me why they aren't

Reply:
Good idea, we'll write up some negatives.

As for this alternative specifically: I hate it. OpenStack's TripleO is solution I am familiar with that wraps well-documented upstream configuration in a layer of productised indirection. It sounds nice in theory, but the experience of it is mostly:

* User frustration when they can't work out how to translate well-documented configuration into a different format
* Unnecessary feature lag when new features haven't been wrapped yet
* Additional maintenance burden of implementing new config options in 2 different places

That isn't to say there aren't advantages, but the negatives are very, very negative for both users and developers.

By contrast, a solution like this which can validate a correct configuration but will otherwise mostly leave it alone is very simple for everybody to understand.